File: androidApp/src/main/java/com/mbta/tid/mbta_app/android/util/RouteStopDirectionExtension.kt
package com.mbta.tid.mbta_app.android.util

import android.content.res.Resources
import com.mbta.tid.mbta_app.android.component.directionNameFormatted
import com.mbta.tid.mbta_app.model.Direction
import com.mbta.tid.mbta_app.model.RouteStopDirection
import com.mbta.tid.mbta_app.model.response.GlobalResponse

data class RouteStopDirectionLabels(val route: String, val stop: String, val direction: String)

fun RouteStopDirection.getLabels(
    global: GlobalResponse?,
    resources: Resources,
): RouteStopDirectionLabels? {
    val lineOrRoute = global?.getLineOrRoute(this.route)
    val stop = global?.getStop(this.stop)

    if (lineOrRoute == null || stop == null) return null

    val routeLabel = lineOrRoute.labelWithModeIfBus(resources)

    val directionLabel =
        resources.getString(
            directionNameFormatted(Direction(this.direction, lineOrRoute.sortRoute))
        )

    return RouteStopDirectionLabels(routeLabel, stop.name, directionLabel)
}

File: iosApp/iosApp/Utils/RouteStopDirectionExtension.swift
//
//  RouteStopDirectionExtension.swift
//  iosApp
//
//  Created by esimon on 8/4/25.
//  Copyright © 2025 MBTA. All rights reserved.
//

import Shared
import SwiftUI

struct RouteStopDirectionLabels {
    let route: String
    let stop: String
    let direction: String
}

extension RouteStopDirection {
    func getLabels(_ global: GlobalResponse?) -> RouteStopDirectionLabels? {
        guard let global,
              let lineOrRoute = global.getLineOrRoute(lineOrRouteId: route),
              let stop = global.getStop(stopId: stop)
        else { return nil }

        let directionLabel = DirectionLabel.directionNameFormatted(.init(
            directionId: direction,
            route: lineOrRoute.sortRoute
        ))

        return .init(route: lineOrRoute.label, stop: stop.name, direction: directionLabel)
    }
}
